ca_tools/scripts/draw_square.py

```diff
@@ -0,0 +1,214 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+
+
+import curses
+import math
+
+import rospy
+from geometry_msgs.msg import Twist, Pose, Quaternion, PoseWithCovariance
+from nav_msgs.msg import Odometry
+from tf.transformations import euler_from_quaternion as efq
+from tf.transformations import quaternion_from_euler as qfe
+
+ANGLE_THRESHOLD = 0.05
+ANGULAR_VEL = 0.3
+LINEAR_VEL = 0.5
+LENGTH_THRESHOLD = 0.05
+GOAL_DIST_THRESHOLD = 0.05
+
+
+class RobotState():
+
+    STATES = [
+    "MOVE_FORWARD",
+    "ROTATE",
+    "STOP"
+    ]
+
+    _current_state = None
+
+    def __init__(self):
+        self._current_state = "STOP"
+
+    def SetState(self,state):
+        "assert a valid state"
+        self._current_state = state
+
+    def GetState(self):
+        return self._current_state
+
+
+class GTSys():
+
+    _pose = Pose()
+
+    def __init__(self):
+        self._sub_gts=rospy.Subscriber('gts', Odometry, self._callback)
+
+    def _callback(self,data):
+        self._pose = data.pose
+
+    def GetPose(self):
+        return self._pose
+
+
+
+class DrawSquare():
+
+
+    _current_pose = PoseWithCovariance()
+    _state = RobotState()
+    _twist = Twist()
+    _next_pose = Pose()
+    _ground_truth = GTSys()
+    _goal_angle = 0
+    _currentGoalReached = False
+    _goal_pose = Pose()
+    _diff_angle = 0
+    _length_diff = 0
+    _square_length = 2
+
+
+    def __init__(self):
+        self._pub_cmd = rospy.Publisher('cmd_vel', Twist, queue_size=10)
+        self._sub_cmd = rospy.Subscriber('cmd_vel', Twist, self._callback)
+        self._hz = rospy.get_param('~hz', 10)
+        self._current_pose = self._ground_truth.GetPose()
+        self._rate = rospy.Rate(self._hz)
+        self._rate.sleep()
+
+
+    def _wrapAngle(self,angle): #ToDo: make beautiful this horrible function
+        while (angle<0):
+            angle+=2.0*math.pi
+        while (angle>2*math.pi):
+            angle-=2.0*math.pi
+        return angle
+
+
+    def _publish(self):
+        self._pub_cmd.publish(self._twist)
+
+    def _callback(self,data):
+        self._twist=data
+
+    def _rotate(self):
+        rospy.loginfo("rotate")
+        self._twist.angular.z = ANGULAR_VEL
+        self._twist.linear.x = 0
+
+        while(self._diff_angle > ANGLE_THRESHOLD):
+            self._publish()
+            self._updateCurrentPose()
+            self._getPoseDiff()
+        self._stop()
+
+
+    def _stop(self):
+        rospy.loginfo("stop")
+        self._twist.angular.z = 0
+        self._twist.linear.x  = 0
+        self._publish()
+
+    def _getPoseDiff(self):
+        q = self._current_pose.pose.orientation
+        euler = efq((q.x,q.y,q.z,q.w))
+        self._current_angle = self._wrapAngle(euler[2])
+
+        diff_x = (self._goal_pose.position.x - self._current_pose.pose.position.x)
+        diff_y = (self._goal_pose.position.y - self._current_pose.pose.position.y)
+
+        self._goal_angle = self._wrapAngle(math.atan2(diff_y,diff_x))
```

Comment:
Check what `atan2` returns and if you need to wrap it.

Reply:
Actually, I'm wrapping around 0 and 2pi, and atan2 return values between -pi and +pi.
Should I wrap around -pi and +pi?

Reply:
Yes, wrapping it between -PI and +PI is recommended.